File: haskell/sum-of-multiples/src/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples) where

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors limit = sum $ filter isMultiple [1.. limit - 1]
  where isMultiple n = any multipleOf factors
          where multipleOf = (== 0) . (n `mod`)

File: java/sum-of-multiples/src/main/java/SumOfMultiples.java
import java.util.stream.IntStream;

public class SumOfMultiples {
	private static final int LIMIT_IS_EXCLUSIVE = 1;
	private final int lastNumber;
	private final int[] factors;

	public SumOfMultiples(final int limit, final int[] factors) {
		this.lastNumber = limit - LIMIT_IS_EXCLUSIVE;
		this.factors = factors;
	}

	public int getSum() {
		return IntStream.of(factors)
			.flatMap(this::multiplies)
			.distinct()
			.sum();
	}

	private IntStream multiplies(final int factor) {
		return IntStream.iterate(factor, n -> n + factor)
			.limit(lastNumber / factor);
	}
}
